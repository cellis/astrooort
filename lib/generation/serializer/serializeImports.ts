import handleImports, { Import } from './handleImports';
import { PascalCase } from './utils';

const getPartial = (imp: Import, importName: string) => {
  let partial = ' ';
  if (imp.partial) {
    partial = `{ ${imp.partial.sort().join(', ')} }`;
  } else {
    partial = PascalCase(importName);
  }
  return partial;
};

export default (model: Superluminal.Model, options?: Superluminal.Args) => {
  const imp0rts = handleImports(model, options?.graphql);

  const sortedImports = Object.keys(imp0rts).sort();

  const topLevelImports = sortedImports
    .filter((i) => imp0rts[i].isModule)
    .map((importName) => {
      const imp = imp0rts[importName];

      return `import ${getPartial(imp, importName)} from '${importName}';`;
    });

  const associationImports = sortedImports
    .filter((i) => !imp0rts[i].isModule)
    .map((importName) => {
      const imp = imp0rts[importName];

      return `import ${getPartial(imp, importName)} from './${PascalCase(
        importName
      )}';`;
    });

  return `${topLevelImports.concat(associationImports).join('\n')}\n`;
};
