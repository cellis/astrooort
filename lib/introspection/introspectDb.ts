import { Introspection } from '@wmfs';
import PgInfo from '@wmfs/pg-info';
import { Client } from 'pg';
import { loadIndexes, RawIndex } from './indexesIntrospection';

export interface IntrospectDB {
  indexes: RawIndex[];
  info: Introspection;
}

export default async (
  client: Client,
  schemas: string[]
): Promise<IntrospectDB> => {
  const info = await PgInfo({
    client,
    schemas,
  });

  const indexes = await loadIndexes(client, schemas.join(','));

  return {
    indexes,
    info,
  };
};
