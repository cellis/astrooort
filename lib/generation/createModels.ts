import { IntrospectDB } from '../introspection/introspectDb';
import createModel from './createModel';

export default function createModels(
  models: Record<string, Superluminal.Model>,
  introspection: IntrospectDB
) {
  const { info, indexes } = introspection;

  for (const [schemaName, schema] of Object.entries(info.schemas)) {
    for (const [tableName, table] of Object.entries(schema.tables)) {
      createModel(tableName, schemaName, table, models, indexes);
    }
  }

  return models;
}
