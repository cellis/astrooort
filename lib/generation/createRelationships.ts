import { IntrospectDB } from '../introspection/introspectDb';
import createRelationship from './createRelationship';
export default function createRelationships(
  models: Superluminal.Models,
  introspection: IntrospectDB,
  associationMapping: Superluminal.AssociationMapping,
  config: Superluminal.Config = {}
) {
  const { info } = introspection;

  if (info.schemas) {
    
    for (const schema of Object.values(info.schemas)) {
      if (schema.tables) {
        for (const [tableName, table] of Object.entries(schema.tables)) {
          createRelationship(
            tableName, 
            table, 
            models, 
            associationMapping, 
            config
            );
        }
      }
    }
  }

  return models;
}
