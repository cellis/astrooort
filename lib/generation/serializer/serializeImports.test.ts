import { Client } from 'pg';
import introspectDb from '../../introspection/introspectDb';
import connectTestDb from '../../test/connectTestDb';
import createModels from '../createModels';
import createRelationships from '../createRelationships';
import serializeImports from './serializeImports';

describe('serializeImports', () => {
  let client: Client;
  let User: Superluminal.Model;
  beforeAll(async (done) => {
    client = await connectTestDb();
    const introspection = await introspectDb(client, ['superluminal']);
    const models: Superluminal.Models = {};
    createModels(models, introspection);
    createRelationships(models, introspection);

    User = models.user;
    done();
  });

  afterAll(() => {
    return client.end();
  });

  it('serializes the imports properly', () => {
    expect(serializeImports(User)).toMatchInlineSnapshot(`
      "import { BaseEntity, Column, Entity, Index, OneToMany, OneToOne } from 'typeorm';
      import { Account } from './Account';
      import { Photo } from './Photo';
      import { Transaction } from './Transaction';
      "
    `);
  });

  it('serializes the graphql imports properly', () => {
    expect(serializeImports(User, { graphql: true })).toMatchSnapshot();
  });
});
