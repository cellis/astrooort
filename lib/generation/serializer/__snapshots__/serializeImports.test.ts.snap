// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeImports serializes the graphql imports properly 1`] = `
"import { Field, ID, ObjectType } from 'type-graphql';
import { BaseEntity, Column, Entity, Index, OneToMany, OneToOne } from 'typeorm';
import { Account } from './Account';
import { Photo } from './Photo';
import { Transaction } from './Transaction';
"
`;
